platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java

```diff
@@ -1305,23 +1394,58 @@ private static <K, V> void purge(final ConcurrentArray<Mutation<K, V>> array, fi
+     * @param index the corresponding index, it's used to look up the newest mutations
+     *              for a key
+     * @param newestVersion the newest version of all mutations in the array
      * @param <K>
      * 		The key type used in the index
      * @param <V>
      * 		The value type referenced by the mutation list
      */
-    private static <K, V> void filterMutations(final ConcurrentArray<Mutation<K, V>> array) {
+    private static <K, V> void filterMutations(
+            final ConcurrentArray<Mutation<K, V>> array,
+            final Map<K, Mutation<K, V>> index,
+            final long newestVersion,
+            final AtomicLong filteredCounter) {
+        filteredCounter.set(0);
         final Consumer<Mutation<K, V>> action = mutation -> {
             // local variable is required because mutation.next can be changed by another thread to null
             // see https://github.com/hashgraph/hedera-services/issues/7046 for the context
             Mutation<K, V> nextMutation = mutation.next;
             if (nextMutation != null) {
+                assert !nextMutation.isFiltered();
                 nextMutation.setFiltered();
+                filteredCounter.incrementAndGet();
+                if (nextMutation.isNew()) {
+                    // nextMutation is to put a new element into a virtual map. The element doesn't
+                    // exist in the data source. If this mutation is filtered, there must be a newer
+                    // mutation for the same key. If that newer mutation has the "deleted" flag, the
+                    // element should never be flushed to disk
+                    final Mutation<K, V> latestMutation = index.get(nextMutation.key);
+                    assert latestMutation != null;
+                    final Mutation<K, V> latestMutationUpToVersion = lookup(latestMutation, newestVersion);
+                    assert latestMutationUpToVersion != null;
+                    if (latestMutationUpToVersion.isDeleted()) {
+                        assert !latestMutationUpToVersion.isFiltered();
+                        // If the latest mutation up to newestVersion is "deleted", and there are no
+                        // newer mutations, the whole entry for the key can be removed from the index.
+                        // It's safe to do so here, as there are no references to copies older than
+                        // newestVersion and there are no mutations in versions newer than newestVersion
+                        if (latestMutation == latestMutationUpToVersion) {
+                            latestMutationUpToVersion.setFiltered();
+                            filteredCounter.incrementAndGet();
+                            index.remove(latestMutation.key);
+                        }
+                    } else {
+                        // Propagate the "new" flag to the newer mutation
+                        latestMutationUpToVersion.setNew();
```

Comment:
Minor: from the code above and the `lookup` behavior, I assume that there may be mutations in between the `nextMutation`/`latestMutation` and the `latestMutationUpToVersion`. The `isNew` flag is going to be propagated from the `nextMutation` to the `latestMutationUpToVersion`, leaving the interim mutations not marked with `isNew`. I've no idea if this may break any logic, today or in the future, but I just want to call this out to check if this has been considered.

Reply:
Check the example in the previous comment. If mutation3 is not "deleted", "new" marker from mutation1 will be propagated to it, but not to version2. This is fine, since version2 (and all subsequent versions) will be removed from the linked list anyway.

Reply:
Got it. Thanks for clarifying!